Exporting query results as Google Spreadsheet

kyoshidajp opened the pull request:
In order to export query results to Google Spreadsheet, calling [IMPORTDATA](https://support.google.com/docs/answer/3093335?hl=en) with CSV format URL from Spreadsheet is the easiest way. However, the way is needed to make access public to allow access from Google.

This feature is export query results by using Google Spreadsheet API and add you as the owner. Also, after you export results, then open the sheet page.

## Steps to use

1. Submit **JSON Key File** which has `https://spreadsheets.google.com/feeds` and `https://www.googleapis.com/auth/drive` scope of Google Spreadsheet  at settings page
1. Run a query and get the result
1. Click **Export button** on query page

## Screenshot

**Export button on query page**

"Download" label may not be an appropriate name.

| Before | After |
| ------- | ------ |
| <img width="200" alt="download_before" src="https://user-images.githubusercontent.com/3317191/36642283-a79d2c6e-1a80-11e8-8535-8e1ff0c462d5.png"> | <img width="208" alt="download_after" src="https://user-images.githubusercontent.com/3317191/36642284-b1b690dc-1a80-11e8-9a7d-48120b16a63f.png"> |

**JSON key file input on settings page**

| Before | After |
| ------- | ------ |
| <img width="300" alt="settings_before" src="https://user-images.githubusercontent.com/3317191/36642336-76d1d340-1a81-11e8-815a-bfa16ff35e9f.png">  | <img width="300" alt="settings_after" src="https://user-images.githubusercontent.com/3317191/36642314-22d8147a-1a81-11e8-9fcc-9be4e48a60ba.png"> |

kyoshidajp commented:
Sorry, I'm going to fix error tests later.

kocsmy commented:
I like this feature, thanks @kyoshidajp 

Question: what happens if you don't have your JSON Key file and your try to hit the "Export as Spreadsheet" button? 

I think it'd be an elegant solution to show a modal with a little message that you'll need to upload your JSON Key File first and link to the Settings (or maybe support article).

The choice of word for Download/Export is indeed a bit confusing. What I'd do is keep it as Download (and we'll figure out something better for this later) but use "Export as Google Spreadsheet" instead to be clear about what's going to happen.

On the Settings page, shouldn't we use "Google Spreadsheet" (instead of Spreadsheet) to be more specific about what is this for?

denisov-vlad commented:
Is it possible to export to Google Spreadsheets without passing JSON key file when the user is registered via OAuth?

kyoshidajp commented:
@kocsmy Thanks. 

> Question: what happens if you don't have your JSON Key file and your try to hit the "Export as Spreadsheet" button?

I didn't consider that. Now, I changed that redirect to settings page with a warning message.

<img width="600" alt="organization settings 2018-02-27 20-23-51" src="https://user-images.githubusercontent.com/3317191/36726161-2d8cae0e-1bfc-11e8-8866-4b546a9adc46.png">

Also, I changed the label on the settings and on the query page.

<img width="256" alt="new querym 2018-02-27 20-19-09" src="https://user-images.githubusercontent.com/3317191/36726280-86ba22c2-1bfc-11e8-9549-1df23bff5c1b.png">

kyoshidajp commented:
@denisov-vlad No. Export by Service Account with API. JSON key file of Google Spreadsheet Account must be set.

kocsmy commented:
Thanks @kyoshidajp that all looks great now!

kyoshidajp commented:
I have resolved the conflict.

bealdav commented:
really interesting feature

arikfr commented:
Hi,

(This is a template message, but I mean every word of it. Also you're welcome to reply)

Thank you for making this contribution. While we couldn't bring it to completion and merge, it's still very much appreciated. 🙇

In the past year the Redash code base gone under massive updates: on the backend we moved to Python 3 & RQ instead of Celery and on the frontend we replaced Angular with React. It's very likely this makes this PR irrelevant without significant changes. :-( 

I'm closing this PR now. But if you're still interested in making it happen, let me know and I will reopen.

Thanks.
